src/uct/tcp/tcp.h

```diff
@@ -27,18 +32,30 @@ typedef struct uct_tcp_am_hdr {
 
 
 /**
- * TCP endpoint
+ * TCP endpoint communication context
  */
-typedef struct uct_tcp_ep {
-    uct_base_ep_t                 super;
-    int                           fd;        /* Socket file descriptor */
-    uint32_t                      events;    /* Current notifications */
-    ucs_queue_head_t              pending_q; /* Pending operations */
+typedef struct uct_tcp_ep_ctx {
     void                          *buf;      /* Partial send/recv data */
     size_t                        length;    /* How much data in the buffer */
     size_t                        offset;    /* Next offset to send/recv */
+} uct_tcp_ep_ctx_t;
+
+
+/**
+ * TCP endpoint
+ */
+struct uct_tcp_ep {
+    uct_base_ep_t                 super;
+    int                           fd;          /* Socket file descriptor */
+    uint32_t                      events;      /* Current notifications */
+    uct_tcp_ep_ctx_t              *tx;         /* TX resources */
+    uct_tcp_ep_ctx_t              *rx;         /* RX resources */
+    uct_tcp_ep_progress_t         progress_tx; /* TX progress enigne */
```

Comment:
can reduce ep size here: declare uint8_t rx and/or tx state, which would be an index in progress funcs array